pkg/probe/fortiswitch_port_stats.go

```diff
@@ -0,0 +1,198 @@
+			"Number of broadcast packets received on the interface",
+			[]string{"vdom", "name", "interface"}, nil,
+		)
+		mRxErr = prometheus.NewDesc(
+			"fortiswitch_port_receive_errors_total",
+			"Number of transmission errors detected on the interface",
+			[]string{"vdom", "name", "interface"}, nil,
+		)
+		mRxDrops = prometheus.NewDesc(
+			"fortiswitch_port_receive_drops_total",
+			"Number of dropped packets detected during transmission on the interface",
+			[]string{"vdom", "name", "interface"}, nil,
+		)
+		mRxOverS = prometheus.NewDesc(
+			"fortiswitch_port_receive_oversized_packets_total",
+			"Number of oversized packets received on the interface",
+			[]string{"vdom", "name", "interface"}, nil,
+		)
+	)
+
+	type portStats struct {
+		TxPackets  float64 `json:"tx-packets"`
+		TxBytes    float64 `json:"tx-bytes"`
+		TxErrors   float64 `json:"tx-errors"`
+		TxMcast    float64 `json:"tx-mcast"`
+		TxUcast    float64 `json:"tx-ucast"`
+		TxBcast    float64 `json:"tx-bcast"`
+		TxDrops    float64 `json:"tx-drops"`
+		TxOversize float64 `json:"tx-oversize"`
+		RxPackets  float64 `json:"rx-packets"`
+		RxBytes    float64 `json:"rx-bytes"`
+		RxErrors   float64 `json:"rx-errors"`
+		RxMcast    float64 `json:"rx-mcast"`
+		RxUcast    float64 `json:"rx-ucast"`
+		RxBcast    float64 `json:"rx-bcast"`
+		RxDrops    float64 `json:"rx-drops"`
+		RxOversize float64 `json:"rx-oversize"`
+	}
+	type portsInfo struct {
+		Interface string
+		Status    string
+		Duplex    string
+		Speed     float64
+		Vlan      string
+	}
+	type swResult struct {
+		Name           string
+		FgPeerIntfName string `json:"fgt_peer_intf_name"`
+		Status         string
+		State          string
+		Connection     string `json:"connecting_from"`
+		VDOM           string
+		Ports          []portsInfo
+		PortStats      map[string]portStats `json:"port_stats"`
+	}
+
+	type swResponse struct {
+		//Results map[string]swResult
+		Results []swResult `json:"results"`
+		//VDOM    string
+	}
+	var r swResponse
+	//var r map[string]swResponse
+	//var r []swResponse
+
+	if err := c.Get("api/v2/monitor/switch-controller/managed-switch", "port_stats=true", &r); err != nil {
+		log.Printf("Error: %v", err)
+		return nil, false
+	}
+	m := []prometheus.Metric{}
+	//	for _, v := range r {
+	for _, swr := range r.Results {
+		swStatus := 0.0
+		if swr.Status == "Connected" {
+			swStatus = 1.0
+		}
+		m = append(m, prometheus.MustNewConstMetric(mSwitchStatus, prometheus.GaugeValue, swStatus, swr.VDOM, swr.Name, swr.FgPeerIntfName, swr.Connection, swr.State))
+
+		for _, pi := range swr.Ports {
+			pStatus := 0.0
+			if pi.Status == "up" {
+				pStatus = 1.0
+			}
+			m = append(m, prometheus.MustNewConstMetric(mPortStatus, prometheus.GaugeValue, pStatus, swr.VDOM, swr.Name, pi.Interface, pi.Vlan, pi.Duplex))
+			m = append(m, prometheus.MustNewConstMetric(mPortSpeed, prometheus.GaugeValue, pi.Speed*1000*1000, swr.VDOM, swr.Name, pi.Interface, pi.Vlan, pi.Duplex))
+
```

Comment:
Having duplex as a label seems a bit off to me. Why not have it as a separate metric?